InfrastructureManager/agents/azure_agent.py

```diff
@@ -0,0 +1,454 @@
+    for network in virtual_networks:
+      result = network_client.virtual_networks.delete(resource_group,
+                                                      network.name)
+      resource_name = 'Virtual Network' + ':' + network.name
+      self.sleep_until_delete_operation_done(result, resource_name)
+
+  def assert_required_parameters(self, parameters, operation):
+    """ Check whether all the parameters required to interact with Azure are
+    present in the provided dict.
+    Args:
+      parameters: A dict containing values necessary to authenticate with the
+        Azure.
+      operation: A str representing the operation for which the parameters
+        should be checked.
+    Raises:
+      AgentConfigurationException: If a required parameter is absent.
+    """
+    # Make sure that the user has set each parameter.
+    for param in self.REQUIRED_CREDENTIALS:
+      if param not in parameters:
+        raise AgentConfigurationException('The required parameter, {0}, was not'
+                                          ' specified.'.format(param))
+
+  def open_connection(self, parameters):
+    """ Connects to Microsoft Azure with the given credentials, creates an
+    authentication token and uses that to get the ServicePrincipalCredentials
+    which is needed to access any resources.
+    Args:
+      parameters: A dict, containing all the parameters necessary to authenticate
+        this user with Azure. We assume that the user has already authorized this
+        account by creating a Service Principal with the appropriate (Contributor)
+        role.
+    Returns:
+      A ServicePrincipalCredentials instance, that can be used to access or
+        create any resources.
+    """
+    app_id = parameters[self.PARAM_APP_ID]
+    app_secret_key = parameters[self.PARAM_APP_SECRET] + "="
+    tenant_id = parameters[self.PARAM_TENANT_ID]
+
+    # Get an Authentication token using ADAL.
+    context = adal.AuthenticationContext(self.AZURE_AUTH_ENDPOINT + tenant_id)
+    token_response = context.acquire_token_with_client_credentials(
+      self.AZURE_RESOURCE_URL, app_id, app_secret_key)
+    token_response.get('accessToken')
+
+    # To access Azure resources for an application, we need a Service Principal
+    # with the accurate role assignment. It can be created using the Azure CLI.
+    credentials = ServicePrincipalCredentials(client_id=app_id,
+                                              secret=app_secret_key,
+                                              tenant=tenant_id)
+    return credentials
+
+  def create_network_interface(self, network_client, interface_name, ip_name,
+                               subnet, parameters):
+    """ Creates the Public IP Address resource and uses that to create the
+    Network Interface.
+    Args:
+      network_client: A NetworkManagementClient instance.
+      interface_name: The name to use for the Network Interface.
+      ip_name: The name to use for the Public IP Address.
```

Comment:
It actually is the name for Public IP Address resource which needs to be created to provide to the Network Interface. The actual IP is the id on that resource and can be retrieved using this name.